monai/data/dataset_calculator.py

```diff
@@ -0,0 +1,128 @@
+                store the metadata of images.
+            num_workers: the maximum number of processes can be used in data loading.
+
+        """
+
+        self.datalist = datalist
+        self.image_key = image_key
+        self.label_key = label_key
+        self.meta_key_postfix = meta_key_postfix
+        self.num_processes = num_processes
+        self.loader = LoadImaged(keys=[image_key, label_key], meta_key_postfix=meta_key_postfix)
+
+    def _run_parallel(self, function):
+        """
+        Parallelly running the function for all data in the datalist.
+
+        """
+        with mp.Pool(processes=self.num_processes) as pool:
+            result = pool.map(function, self.datalist)
+            return result
+
+    def _load_spacing(self, path_dict: Dict):
+        """
+        Load spacing from a data's dictionary. Assume that the original image file has `pixdim`
+        in its metadata.
+
+        """
+        data = self.loader(path_dict)
+        meta_key = "{}_{}".format(self.image_key, self.meta_key_postfix)
+        spacing = data[meta_key]["pixdim"][1:4].tolist()
+
+        return spacing
+
+    def _get_target_spacing(self, anisotropic_threshold: int = 3, percentile: float = 10.0):
+        """
+        Calculate the target spacing according to all spacings.
+        If the target spacing is very anisotropic,
+        decrease the spacing value of the maximum axis according to percentile.
+
+        """
+        spacing = self._run_parallel(self._load_spacing)
+        spacing = np.array(spacing)
+        target_spacing = np.median(spacing, axis=0)
+        if max(target_spacing) / min(target_spacing) >= anisotropic_threshold:
+            largest_axis = np.argmax(target_spacing)
+            target_spacing[largest_axis] = np.percentile(spacing[:, largest_axis], percentile)
+
+        output = list(target_spacing)
+        output = [round(value, 2) for value in output]
+
+        return tuple(output)
+
+    def _load_intensity(self, path_dict: Dict):
+        """
+        Load intensity from a data's dictionary.
+
+        """
+        data = self.loader(path_dict)
+        image = data[self.image_key]
+        foreground_idx = np.where(data[self.label_key] > 0)
+
+        return image[foreground_idx].tolist()
+
+    def _get_intensity_stats(self, lower: float = 0.5, upper: float = 99.5):
+        """
+        Calculate min, max, mean and std of all intensities. The minimal and maximum
+        values will be processed according to the provided percentiles.
+
+        """
+        intensity = self._run_parallel(self._load_intensity)
+        intensity = np.array(list(itertools.chain.from_iterable(intensity)))
+        min_value, max_value = np.percentile(intensity, [lower, upper])
+        mean_value, std_value = np.mean(intensity), np.std(intensity)
+        output = [min_value, max_value, mean_value, std_value]
+        output = [round(value, 2) for value in output]
+
+        return tuple(output)
```

Comment:
All these methods begin with underscore, are they meant to be called by another method (otherwise this implies they're meant to be private)?

Reply:
Fixed.